### PR TITLE
[SYCL] Report an error when SYCL_DEVICE_TYPE is set incorrectly.

### PR DESCRIPTION
**Pull request by glyons-intel**
Signed-off-by: Gail Lyons <gail.lyons@intel.com>

### CODE REVIEW

File: sycl/test/basic_tests/diagnostics/device-check.cpp [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+// RUN: env SYCL_DEVICE_TYPE=host %t.out
+// RUN: env SYCL_DEVICE_TYPE=CPU %t.out
+// RUN: env SYCL_DEVICE_TYPE=GPU %t.out
+// RUN: env SYCL_DEVICE_TYPE=ACC %t.out
+// RUN: env SYCL_DEVICE_TYPE=HOST %t.out
+// RUN: env SYCL_DEVICE_TYPE=Cpu %t.out
+// RUN: env SYCL_DEVICE_TYPE=Gpu %t.out
+// RUN: env SYCL_DEVICE_TYPE=Acc %t.out
+// RUN: env SYCL_DEVICE_TYPE=Host %t.out
+
```
Comment (AGindinson) [inline sycl/test/basic_tests/diagnostics/device-check.cpp]: I believe the separate negative test can be replaced with an addition like:
```suggestion
// RUN: not %clangxx -fsycl %s -o %t.out 2>&1 | FileCheck %s
// CHECK: SYCL_DEVICE_TYPE is not available.  Must be GPU, CPU, ACC or HOST
```

File: sycl/source/detail/force_device.cpp [hunk 2]
```diff
@@ -21,18 +22,22 @@ bool match_types(const info::device_type &l, const info::device_type &r) {
...
     }
-    if (string_class(val) == "ACC") {
+    if (type == "acc") {
       return info::device_type::accelerator;
     }
-    if (string_class(val) == "HOST") {
+    if (type == "host") {
       return info::device_type::host;
     }
+    throw cl::sycl::runtime_error("SYCL_DEVICE_TYPE is not available.  Must be GPU, CPU, ACC or HOST");
```
Comment (AGindinson) [inline sycl/source/detail/force_device.cpp]:
```suggestion
    throw cl::sycl::runtime_error("SYCL_DEVICE_TYPE is not available. Must be GPU, CPU, ACC or HOST");
```

File: sycl/source/detail/force_device.cpp [hunk 3]
```diff
@@ -21,18 +22,22 @@ bool match_types(const info::device_type &l, const info::device_type &r) {
 
 info::device_type get_forced_type() {
   if (const char *val = std::getenv("SYCL_DEVICE_TYPE")) {
-    if (string_class(val) == "CPU") {
+    std::string type(val);
+    std::transform(type.begin(), type.end(), type.begin(), ::tolower);
```
Comment (AGindinson) [inline sycl/source/detail/force_device.cpp]: Or `::toupper`? :) Just to "retain the spirit", I guess

File: sycl/test/basic_tests/diagnostics/device-check.cpp (hunk 1 above)
Comment (glyons-intel) [reply sycl/test/basic_tests/diagnostics/device-check.cpp]: Hi Artem,

Thanks for the code review comments. 

Personally, I prefer several smaller tests to one larger test.  I particular, I separate negative tests from positive tests.  Is there a specific "rule of thumb" about this in LLVM?  If Icould keep two test cases, I would prefer that. 
I appreciate your suggestion on using FileCheck.  That had not occurred to me.

File: sycl/source/detail/force_device.cpp (hunk 2 above)
Comment (glyons-intel) [reply sycl/source/detail/force_device.cpp]: This was intentional.  I was taught that it is 2 spaces after a period.  Do you do it differently?

File: sycl/source/detail/force_device.cpp (hunk 3 above)
Comment (glyons-intel) [reply sycl/source/detail/force_device.cpp]: Obviously, I prefer "tolower()".

File: sycl/test/basic_tests/diagnostics/device-check.cpp (hunk 1 above)
Comment (AGindinson) [reply sycl/test/basic_tests/diagnostics/device-check.cpp]: @glyons-intel, feel free to treat my feedback as optional, for I have no strong objections to the current patch. https://llvm.org/docs/TestingGuide.html#writing-new-regression-tests was the inspiration for this particular comment, but again, "rule of thumb" isn't really implied there.

File: sycl/source/detail/force_device.cpp (hunk 2 above)
Comment (AGindinson) [reply sycl/source/detail/force_device.cpp]: Reflex was my only source of information for this comment, so please ignore the suggestion if there's a deeper point/guideline I'm missing.

File: sycl/test/basic_tests/diagnostics/device-check.cpp (hunk 1 above)
Comment (bader) [reply sycl/test/basic_tests/diagnostics/device-check.cpp]: This suggestion makes sense to me.
1. It removes duplication - device-check.cpp and device-type.cpp source code is the same.
2. It saves on compilation - one of the goals of testing system is to be as fast as possible, so saving compilation is good thing to do.

File: sycl/source/detail/force_device.cpp [hunk 4]
```diff
@@ -10,6 +10,7 @@
 #include <CL/sycl/info/info_desc.hpp>
 #include <CL/sycl/stl.hpp>
 #include <cstdlib>
+#include <algorithm>
```
Comment (romanovvlad) [inline sycl/source/detail/force_device.cpp]: Please, sort header and separate standard header by an empty line.

File: sycl/source/detail/force_device.cpp (hunk 2 above)
Comment (romanovvlad) [inline sycl/source/detail/force_device.cpp]: "SYCL_DEVICE_TYPE is not available." sounds confusing, what about:
"SYCL_DEVICE_TYPE value is not recognized." ?

File: sycl/source/detail/force_device.cpp (hunk 4 above)
Comment (glyons-intel) [reply sycl/source/detail/force_device.cpp]: Done.

File: sycl/source/detail/force_device.cpp (hunk 2 above)
Comment (glyons-intel) [reply sycl/source/detail/force_device.cpp]: Done.

**Comment by glyons-intel** [comment]
Artem & Alexey, 
I did combine both the positive and negative tests into one file, as requested.
Gail

File: sycl/source/detail/force_device.cpp (hunk 4 above)
Comment (bader) [reply sycl/source/detail/force_device.cpp]: I think `#include <algorithm>` should go before `#include <cstdlib>`.

File: sycl/source/detail/force_device.cpp [hunk 5]
```diff
@@ -21,18 +23,22 @@ bool match_types(const info::device_type &l, const info::device_type &r) {
 
 info::device_type get_forced_type() {
   if (const char *val = std::getenv("SYCL_DEVICE_TYPE")) {
-    if (string_class(val) == "CPU") {
+    std::string type(val);
+    std::transform(type.begin(), type.end(), type.begin(), ::tolower);
+
```
Comment (bader) [inline sycl/source/detail/force_device.cpp]: Please, apply clang-format to the patch.

Comment (glyons-intel) [reply sycl/source/detail/force_device.cpp]: Done.

**Comment by AlexeySachkov** [comment]
> Interesting... I don't expect clang-format to indent namespace here.
> I thought that it's fixed by [intel#1121](https://github.com/intel/llvm/pull/1121).
> @AlexeySachkov, any ideas what could be the problem? Old clang-format version?

Most likely. Corresponding feature was added to clang-format just ~8 month ago: https://reviews.llvm.org/D37813

**Comment by glyons-intel** [comment]
This is the version of clang-format that I am using:

% clang-format --version
clang-format version 11.0.0 (https://github.com/glyons-intel/llvm 41b73bec8453bc22e03dccb27be9472072770290)

I invoked it like this: 
clang-format --style=llvm force_device.cpp > new_force_device.cpp

**Comment by bader** [comment]
> This is the version of clang-format that I am using:
> 
> % clang-format --version
> clang-format version 11.0.0 (https://github.com/glyons-intel/llvm 41b73bec8453bc22e03dccb27be9472072770290)
> 
> I invoked it like this:
> clang-format --style=llvm force_device.cpp > new_force_device.cpp

Could you remove `--style=llvm`, please?
I think it's the problem. It disables the configuration we did the project here: https://github.com/intel/llvm/blob/sycl/sycl/.clang-format.

**Comment by glyons-intel** [comment]
Yes, that made a difference.
I think I read about using --style=llvm in the standard clang-format documentation.
I will update my patch.

**Pull request merged**